### PR TITLE
[1.0] Add "duotone" option to gatsby-plugin-sharp

### PR DESCRIPTION
**Pull request by fk**
This adds a "duotone" option (see [I][1], [II][2], [III][3]) to `gatsby-plugin-sharp`.

It also changes the Base64 images generated by `responsiveResolution` and `responsiveSizes` to inherit `grayscale` and `duotone` effects applied to their respective source images.

Given two hex colors `shadow` and `highlight` defining start and end color of the duotone gradient, e.g.

```
responsiveResolution(
  width: 800,
  duotone: {
    highlight: "#f00e2e",
    shadow: "#192550"
  }
) {
  src
  srcSet
  base64
}
```

the source image colors will be converted to match a gradient color chosen based on each pixel's [relative luminance][4]. Logic is borrowed from [react-duotone][5].

[1]: https://alistapart.com/article/finessing-fecolormatrix
[2]: http://blog.72lions.com/blog/2015/7/7/duotone-in-js
[3]: https://ines.io/blog/dynamic-duotone-svg-jade
[4]: https://en.wikipedia.org/wiki/Relative_luminance
[5]: https://github.com/nagelflorian/react-duotone

### CODE REVIEW

**Comment by KyleAMathews** [comment]
Yeah! Looks amazing!

Canihaz an example site too?

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 528821a4a014f43362126871fd0d80d6e883aca1

https://app.netlify.com/sites/using-postcss-sass/deploys/592808956f4c506787132e81

**Comment by fk** [comment]
The cheez is live on e. g. https://www.yisela.com/your-bastardized-version-scrum-actually-harming/. Still have to catch up with alpha16 though.

I also tested with the defacto [reference image](https://jmperezperez.com/assets/images/posts/fecolormatrix-kanye-west.jpg) for all the duotone stuff linked above:

fecolormatrix-kanye-west

High contrast images like above work best, but they don't need to be necessarily grayscaled already before applying the duotone effect. Depending on the source image, the quality needs to be set quite high to get rid of artifacts. For Kanye I did this

```
responsiveSizes(
  width: 756,
  quality: 95,
  duotone: {
    highlight: "#f00e2e",
    shadow: "#192550"
  }
) {
  src
  srcSet
  base64
}
```

and got this, which matches the outcome of the various examples linked above:

77e6f7a658028a2953e5ff1f7a849b32-maxwidth 756 duotone 5bhighlight 5d 23f00e2e duotone 5bshadow 5d 23192550 quality 100 toformat jpg jpegprogressive true width 1024

**Comment by fk** [comment]
Ah damn, now I nearly forgot one important thing: The duotone images work fine in development and on the (Netlify) production server. When previewing the build locally however (using `gatsby serve-build`), they get served as JSON. 🙀 

Will grab you a screenshot once I have upgraded to alpha16.

**Comment by KyleAMathews** [comment]
@fk you want to pull in @0x80's recent changes to see if it fixes the 404 error you're seeing?

**Comment by fk** [comment]
@KyleAMathews I expect it will – will check in a few minutes, just rebased on top of latest 1.0 and started working on the example site (getting a little annoyed by that Kanye image ;-)):

image

image

File: packages/gatsby-plugin-sharp/src/index.js [hunk 1]
```diff
@@ -68,30 +57,49 @@ const processFile = (file, jobs, cb) => {
...
-                quality: `${args.quality}-${Math.min(args.quality + 25, 100)}`, // e.g. 40-65
-              }),
-            ],
-          })
-          .then(imageminBuffer => {
-            fs.writeFile(job.outputPath, imageminBuffer, () => {
-              bar.tick()
-              job.outsideResolve()
+    // duotone and write to file
+    duotone(
```
Comment (KyleAMathews) [inline packages/gatsby-plugin-sharp/src/index.js]: It'd be clearer to check if duotone is set as an option here vs. inside duotone.js

Comment (fk) [inline packages/gatsby-plugin-sharp/src/index.js]: Agreed. Struggled a lot here waiting for the async function, but couldn't come up with a better solution. Eager to find one though – maybe you have a pointer for me?!

Comment (KyleAMathews) [inline packages/gatsby-plugin-sharp/src/index.js]: Normal if/else works with async/await

Comment (fk) [inline packages/gatsby-plugin-sharp/src/index.js]: Fixed … thanks Kyle! 🙏  😉

**Comment by KyleAMathews** [comment]
This is looking so good! Love the new image processing example site! Let me pull it in now and make more improvements in future PRs.

Thanks for the new Duotone option + docs + example site!

**Comment by fk** [comment]
Small follow up to the [issue regarding duotone images](https://github.com/gatsbyjs/gatsby/pull/1047#issuecomment-304258255) (with percent-encoded reserved characters in the filename) and locally previewing the build: Fixed as expected, just tested with alpha19. ✌️

**Pull request merged**